server/sonar-ce/src/main/java/org/sonar/ce/container/ComputeEngineContainerImpl.java

```diff
@@ -646,6 +199,473 @@ protected ComponentContainer getComponentContainer() {
+    return new Object[] {
+      ComputeEngineSettings.class,
+      new SonarQubeVersion(apiVersion),
+      SonarRuntimeImpl.forSonarQube(ApiVersion.load(System2.INSTANCE), SonarQubeSide.COMPUTE_ENGINE),
+      ServerImpl.class,
+      UuidFactoryImpl.INSTANCE,
+      // no EmbeddedDatabaseFactory.class, creating H2 DB if responsibility of WebServer
+      DefaultDatabase.class,
+      DatabaseChecker.class,
+      // must instantiate deprecated class in 5.2 and only this one (and not its replacement)
+      // to avoid having two SqlSessionFactory instances
+      org.sonar.core.persistence.MyBatis.class,
+      DatabaseServerCompatibility.class,
+      DatabaseVersion.class,
+      PurgeProfiler.class,
+      DefaultServerFileSystem.class,
+      // no TempFolderCleaner.class, responsibility of Web Server
+      new TempFolderProvider(),
+      System2.INSTANCE,
+
+      // user session
+      CeUserSession.class,
+
+      // DB
+      DaoModule.class,
+      // DbClient.class, replaced by CeDbClient to use ReadOnlyPropertiesDao instead of PropertiesDao
+      ReadOnlyPropertiesDao.class,
+      DbClient.class,
+      // MigrationStepModule.class, DB maintenance, responsibility of Web Server
+
+      // Elasticsearch
+      EsSearchModule.class,
+
+      // rules/qprofiles
+      RuleIndex.class,
+      ActiveRuleIndex.class,
+
+      // issues
+      IssueIndex.class,
+
+      // Classes kept for backward compatibility of plugins/libs (like sonar-license) that are directly calling classes from the core
+      // org.sonar.core.properties.PropertiesDao.class, replaced by ReadOnlyPropertiesDao (declared above) which is a ReadOnly
+      // implementation
+    };
+  }
+
+  private static Object[] level2Components() {
+    return new Object[] {
+      // add ReadOnlyPropertiesDao at level2 again so that it shadows PropertiesDao
+      ReadOnlyPropertiesDao.class,
+      DefaultServerUpgradeStatus.class,
+      // no DatabaseMigrator.class, responsibility of Web Server
+
+      // plugins
+      PluginClassloaderFactory.class,
+      CePluginJarExploder.class,
+      PluginLoader.class,
+      CePluginRepository.class,
+      InstalledPluginReferentialFactory.class,
+      ComputeEngineExtensionInstaller.class,
+
+      // depends on plugins
+      // RailsAppsDeployer.class,
+      // JRubyI18n.class,
+      DefaultI18n.class, // used by RuleI18nManager
+      RuleI18nManager.class, // used by DebtRulesXMLImporter
+      Durations.class, // used in Web Services and DebtCalculator
+    };
+  }
+
+  private static Object[] level3Components() {
+    return new Object[] {
+      PersistentSettings.class,
+      // ServerMetadataPersister.class, server id is the responsibility of Web Server
+      // DefaultHttpDownloader.class, does not make sense to use it from Compute Engine
+      UriReader.class,
+      // ServerIdGenerator.class, server id is the responsibility of Web Server
+    };
+  }
+
+  private static Object[] level4Components() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This method has 360 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS138)